### PR TITLE
Various noise gate, scope, and audio stats improvements

### PR DESCRIPTION
**Pull request by ZappoMan**
- Fixes a bug in the noise gate for determining if it was opened on the last frame or not - previously the noise gate would over report that it had just closed.
- Adds signals for scripts to know when the noise gate opens/closes
- Adds scripting interface to the Audio scope to hide/show it and to pause/play it
- Adds fade in/fade out the the noise gate so that we don't get harsh clipping when opening/closing the gate.
- Adds a simple Tablet/Toolbar app that supports hiding/showing the scope and demonstrating these features.
- Adds new audio related stats to the "stats" overlay.
    - Audio Mixer inbound/outbound stats for bandwidth and PPS
    - Silent vs Mic/Audio inbound/outbound PPS stats
    - Which codec is in use
    - Noise gate open/closed

### CODE REVIEW

File: libraries/audio-client/src/AudioNoiseGate.cpp [hunk 1]
```diff
@@ -141,17 +141,37 @@ void AudioNoiseGate::gateSamples(int16_t* samples, int numSamples) {
...
+    _closedInLastFrame = false;
+    _openedInLastFrame = false;
+
     if (samplesOverNoiseGate > NOISE_GATE_WIDTH) {
+        _openedInLastFrame = !_isOpen;
         _isOpen = true;
         _framesToClose = NOISE_GATE_CLOSE_FRAME_DELAY;
     } else {
         if (--_framesToClose == 0) {
-            _closedInLastFrame = !_isOpen;
```
Comment (ZappoMan) [inline libraries/audio-client/src/AudioNoiseGate.cpp]: This was the source of a bug that caused the noise gate to constantly report that it had "just closed" resulting in the client not properly sending silent packets.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 5f3ee2e98f1285f64d392b07d8832caf5b9aa95b available at: https://gist.github.com/5c8eaba233df7b771b691148a96f0e70<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9862-5f3ee2e.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9862-5f3ee2e.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 7eb4b8f5f73f8c39c5cdc5219d96c2555a196c25 available at: https://gist.github.com/5c8eaba233df7b771b691148a96f0e70<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9862-7eb4b8f.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9862-7eb4b8f.dmg

File: assignment-client/src/avatars/AvatarMixer.cpp [hunk 2]
```diff
@@ -37,7 +37,6 @@ const QString AVATAR_MIXER_LOGGING_NAME = "avatar-mixer";
 
 // FIXME - what we'd actually like to do is send to users at ~50% of their present rate down to 30hz. Assume 90 for now.
 const int AVATAR_MIXER_BROADCAST_FRAMES_PER_SECOND = 45;
-const unsigned int AVATAR_DATA_SEND_INTERVAL_MSECS = (1.0f / (float) AVATAR_MIXER_BROADCAST_FRAMES_PER_SECOND) * 1000;
```
Comment (ZappoMan) [inline assignment-client/src/avatars/AvatarMixer.cpp]: this variable is unused, causes a warning on mac

**Comment by ZappoMan** [comment]
# TEST PLAN

## AUDIO-SCOPE-01 Test
Objective: Verify that audio scope renders mic and server audio.
Precondition: Enter any domain, but preferably one with limited audio.
Step 1: Run scripts/system/audioScope.js
Step 2: Enable the audio scope.
Step 3: Enable Server Loopback
Step 4: Make sure Audio Scope is not paused.
Step 5: Speak and verify that a green line draws your mic audio, and slightly delayed is a blue line with the echo of your server loopback.
Step 6: Both lines should look basically the same.

## AUDIO-SCOPE-02 Test
Objective: Verify that audio scope pauses and resumes correctly.
Precondition: Complete AUDIO-SCOPE-01 Test verify that the audio scope is rendering.
Step 1: Run scripts/system/audioScope.js
Step 2: Enable the audio scope.
Step 3: Enable Server Loopback
Step 4: Verify that the Audio scope pauses, using the menu, the toggle button on the toolbar/tablet, and the menu hotkey.
Step 5: Verify that the Audio scope unpauses. 

## AUDIO-SCOPE-03 Test
Objective: Verify that audio scope scripting interface
Precondition: None
Step 1: Run interface.
Step 2: Open Edit > Console.
Step 3: Enter the following commands:
AudioScope.setVisible(true); // verify that the audio scope is enabled and renders
AudioScope.setVisible(false); // verify that the audio scope is hidden
AudioScope.setPause(true); // verify that the audio scope pauses and freezes the rendered audio
AudioScope.setPause(false); // verify that the audio scope is not paused and shows your rendered audio scrolling to the left

## NOISE-GATE-01 Test
Objective: Verify that noise gate opens and closes as expected.
Precondition: Enter any domain, but preferably one with limited audio, such that you are the only person in the domain. Make sure the Audio Noise Reduction is enabled.
Step 1: Run scripts/system/audioScope.js
Step 2: Enable the audio scope, and enable "auto pause".
Step 3: Try being quite and speaking alternately, verify that the audio scope visualization automatically pauses when you are mostly quite, and starts when you make noise.

## NOISE-GATE-02 Test
Objective: Verify that when the noise gate closes and your client sends silent packets that there is no distortion in the echoed audio.
Precondition: Enter any domain, make sure that the domain is totally silent other than your own client. Make sure the Audio Noise Reduction is enabled.
Step 1: Run scripts/system/audioScope.js
Step 2: Enable the audio scope, and enable "auto pause".
Step 3: Enable Server Loopback
Step 4: Try being quite and speaking alternately, verify that the audio scope visualization automatically pauses when you are mostly quite, and starts when you make noise.
Step 5: Verify that after being quiet and beginning to speak, that the echoed audio is a close match to the outbound audio wave form. It should not contain significant distortions. You should listen closely for things like apparent "type writer" sounds or other distortions, and you should inspect the paused audio scope wave form to make sure it looks like a match.

## NOISE-GATE-03 Test
Objective: Verify that reports of noise gate opening and closing are always in sequential pairs, that we don't report that the noise gate closes more than once in a row.
Precondition: Enter any domain, make sure that the domain is totally silent other than your own client. Make sure the Audio Noise Reduction is enabled.
Step 1: Run interface.
Step 2: Open Edit > Console.
Step 3: Enter the following commands:
Audio.noiseGateOpened.connect(function() { print("opened"); }); Audio.noiseGateClosed.connect(function() { print("closed"); });
Step 4: Verify that in the log you see "opened" followed by "closed" and then repeated each time the noise gate opens or closes.

**Comment by hifi-gustavo** [comment]
Build console output for commit: aa65f03509f9a56f5e060c1e95025283b78b377f available at: https://gist.github.com/5c8eaba233df7b771b691148a96f0e70<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9862-aa65f03.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9862-aa65f03.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 23e47f2fdc8de48290a0a7dd8f0847213b2e121c available at: https://gist.github.com/5c8eaba233df7b771b691148a96f0e70<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9862-23e47f2.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9862-23e47f2.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 6ce50269ac62278f13d1d4253b0427b575234766 available at: https://gist.github.com/5c8eaba233df7b771b691148a96f0e70<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9862-6ce5026.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9862-6ce5026.dmg

File: libraries/shared/src/shared/RateCounter.h [hunk 3]
```diff
@@ -24,29 +24,34 @@ class RateCounter {
...
-            _rate = roundf(currentCount / intervalSeconds * _scale) / _scale;
-            _start = now;
-            _count = 0;
-        };
+        checkRate();
         _count += count;
     }
 
-    float rate() const { return _rate; }
+    float rate() const { checkRate(); return _rate; }
```
Comment (ZappoMan) [inline libraries/shared/src/shared/RateCounter.h]: We need to check for rate state changes (like completing an interval) when you request the rate, or else you get bad data when the rate actually stops getting updated.

**Comment by hifi-gustavo** [comment]
Build console output for commit: a38965d628b09dfe179c1dccef4c56059a8e37fb available at: https://gist.github.com/5c8eaba233df7b771b691148a96f0e70<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9862-a38965d.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9862-a38965d.dmg

**Comment by OmegaHeron** [comment]
Windows 10, QA Pass

Test | Result
---|---
AUDIO-SCOPE-01 Test | Pass
AUDIO-SCOPE-02 Test | Pass
AUDIO-SCOPE-03 Test | Pass
NOISE-GATE-01 Test | Pass
NOISE-GATE-02 Test | Pass *See notes*
NOISE-GATE-03 Test | Pass
AUDIO-BANDWIDTH-01 Test | Pass

---

Notes:
tldr; It seems perfectly acceptable if evaluation is less than hypercritical.

"You should listen closely for things like apparent "type writer" sounds or other distortions"

I'm translating that to be the artifact occurring when gate opens and you're trying to ease in versus having a harsh transient.  Compared to previous iterations, this code is much better, but, it depends heavily upon several things like speaker's voice, microphone characteristics and room's background noise level.  Overall, I heard no objectionable artifacts, but to be thorough the following should be known results.

Room with minimal background noise - little to no tick/click sound on noise gate open.

Room with heavy background noise (as in things like noisy fans) - some very minor tick/click on noise gate open, but not, in my opinion, of major concern considering such a high white noise background.

Microphone 1 - Omnidirectional pattern
As expected, worst overall for high noise environment with noticeable (but not objectionable) ticks/clicks on some noise gate opens, highly dependent upon sound that opens gate.  Things like keyboard clicks or mouse button clicks or speaking hard T sound like Tick, tink, tock.

Microphone 2 - Cardioid pattern
 Substantially better, as expected, as less input is received from sources other than speaker's voice.

Modifying frequency response of microphone (I.e. applying a low pass filter) offers some potential to help.  Noticed that my cheap consumer grade microphone(s) performed better for tink/click sounds vs any in my collection of broadcast studio microphones.  Mics with little roll off in sensitivity at the high end (15+ KHz) produced more artifacts than those that roll off around 8-10 KHz.

**Comment by ZappoMan** [comment]
@OmegaHeron on the case of NOISE-GATE-02 test. The following phrase is important: "Verify that after being quiet and beginning to speak, that the echoed audio is a close match to the outbound audio wave form. It should not contain significant distortions. " -- which is to say, there was a bug that we have now fixed... which would cause a very significant distortion. It was caused by the last bit of audio (when the noise gate closed) getting pushed into the audio of when it gets opened. If you looked closely at the scope you could actually see this wave form... most people described that sound as a typewriter like sound... it is indeed **very tricky** to hear it, and it sounds a lot like just a normal noise gate opening. If you can think of a better test and or better way to describe this, I would gladly include it in the test plan.

**Comment by OmegaHeron** [comment]
@ZappoMan - It's likely described as well as it can be.  If it's broken there's no missing it as long as the crucial *you must have noise reduction enabled* is respected.

I spent a few years in the 80s as an engineer at AM/FM radio stations where, besides keeping transmitters alive that were from the 1960s and falling apart, I had to setup mics for the announcers.  In the pre DSP era that involved a lot of inserting foam in mic cases and beating into humans that you don't breathe, sneeze, cough, yell or make any other unauthorized sounds into the damn microphone.  This is also why I have a couple boxes filled with high end studio mics of the 1960s/1970s that were retired and saved from living in a landfill as we replaced.

File: libraries/audio-client/src/AudioNoiseGate.cpp [hunk 4]
```diff
@@ -141,17 +141,37 @@ void AudioNoiseGate::gateSamples(int16_t* samples, int numSamples) {
...
+            _closedInLastFrame = _isOpen;
             _isOpen = false;
         }
     }
     if (!_isOpen) {
-        memset(samples, 0, numSamples * sizeof(int16_t));
+        if (_closedInLastFrame) {
+            // would be nice to do a little crossfade to silence
+            for (int i = 0; i < numSamples; i++) {
+                float fadedSample = (1.0f - (float)i / (float)numSamples) * (float)samples[i];
```
Comment (kencooke) [inline libraries/audio-client/src/AudioNoiseGate.cpp]: Dividing by numSamples every sample is kinda expensive. A faster way is to compute slope = (1.0f/numSamples) once and multiply in the inner loop.

File: libraries/audio-client/src/AudioNoiseGate.cpp [hunk 5]
```diff
@@ -141,17 +141,37 @@ void AudioNoiseGate::gateSamples(int16_t* samples, int numSamples) {
...
     } else {
         if (--_framesToClose == 0) {
-            _closedInLastFrame = !_isOpen;
+            _closedInLastFrame = _isOpen;
             _isOpen = false;
         }
     }
     if (!_isOpen) {
-        memset(samples, 0, numSamples * sizeof(int16_t));
+        if (_closedInLastFrame) {
```
Comment (kencooke) [inline libraries/audio-client/src/AudioNoiseGate.cpp]: The negative logic here is a bit confusing. Maybe a comment that explains the first closed frame gets faded to silence, and following ones are muted.

File: libraries/audio-client/src/AudioNoiseGate.cpp [hunk 6]
```diff
@@ -141,17 +141,37 @@ void AudioNoiseGate::gateSamples(int16_t* samples, int numSamples) {
...
+        } else {
+            memset(samples, 0, numSamples * sizeof(int16_t));
+        }
         _lastLoudness = 0;
     }
+
+    if (_openedInLastFrame) {
+        // would be nice to do a little crossfade from silence
+        for (int i = 0; i < numSamples; i++) {
+            float fadedSample = ((float)i / (float)numSamples) * (float)samples[i];
```
Comment (kencooke) [inline libraries/audio-client/src/AudioNoiseGate.cpp]: See above about optimizing the division

File: libraries/audio-client/src/AudioNoiseGate.cpp [hunk 7]
```diff
@@ -141,17 +141,37 @@ void AudioNoiseGate::gateSamples(int16_t* samples, int numSamples) {
...
+                float fadedSample = (1.0f - (float)i / (float)numSamples) * (float)samples[i];
+                samples[i] = (int16_t)fadedSample;
+            }
+        } else {
+            memset(samples, 0, numSamples * sizeof(int16_t));
+        }
         _lastLoudness = 0;
     }
+
+    if (_openedInLastFrame) {
```
Comment (kencooke) [inline libraries/audio-client/src/AudioNoiseGate.cpp]: A fade-in with no lookahead will cause artifacts when the gate is opened by a transient. If the transient occurs at end-of-frame, it will be preserved, but at beginning of frame, it will reshaped/removed. 

Solving this without adding a full frame (10ms) of latency is kinda tricky.

Comment (ZappoMan) [inline libraries/audio-client/src/AudioNoiseGate.cpp]: are you suggesting a fix? note: current behavior is to open the gate and pass through the original buffer... doesn't that current code also suffer from the same problem you describe?

File: libraries/audio-client/src/AudioClient.cpp [hunk 8]
```diff
@@ -1041,6 +1049,13 @@ void AudioClient::handleAudioInput() {
 
             emit inputReceived({ reinterpret_cast<char*>(networkAudioSamples), numNetworkBytes });
 
+            if (_inputGate.openedInLastFrame()) {
+                emit noiseGateOpened();
+            }
+            if (_inputGate.closedInLastFrame()) {
+                emit noiseGateClosed();
+            }
```
Comment (AndrewMeadows) [inline libraries/audio-client/src/AudioClient.cpp]: _openedInLastFrame and _closedInLastFrame are mutually exclusive: only one can be true at any time, so the code would still be correct if the second check was an `else if`.

Comment (ZappoMan) [inline libraries/audio-client/src/AudioClient.cpp]: will fix
